codex/stores/repostore.nim

```diff
@@ -297,131 +432,64 @@ method putBlock*(
   logScope:
     cid = blk.cid
 
-  if blk.isEmpty:
-    trace "Empty block, ignoring"
-    return success()
+  let expiry = self.clock.now() + (ttl |? self.blockTtl).seconds
 
-  without key =? makePrefixKey(self.postFixLen, blk.cid), err:
-    trace "Error getting key from provider", err = err.msg
+  without res =? await self.storeBlock(blk, expiry), err:
     return failure(err)
 
-  if await key in self.repoDs:
-    trace "Block already in store", cid = blk.cid
-    return success()
-
-  if (self.totalUsed + blk.data.len.uint) > self.quotaMaxBytes:
-    error "Cannot store block, quota used!", used = self.totalUsed
-    return failure(
-      newException(QuotaUsedError, "Cannot store block, quota used!"))
-
-  trace "Storing block with key", key
-
-  var
-    batch: seq[BatchEntry]
-
-  let
-    used = self.quotaUsedBytes + blk.data.len.uint
-
-  if err =? (await self.repoDs.put(key, blk.data)).errorOption:
-    trace "Error storing block", err = err.msg
-    return failure(err)
-
-  trace "Updating quota", used
-  batch.add((QuotaUsedKey, @(used.uint64.toBytesBE)))
-
-  without blockExpEntry =? self.getBlockExpirationEntry(blk.cid, ttl), err:
-    trace "Unable to create block expiration metadata key", err = err.msg
-    return failure(err)
-  batch.add(blockExpEntry)
-
-  if err =? (await self.metaDs.put(batch)).errorOption:
-    trace "Error updating quota bytes", err = err.msg
-
-    if err =? (await self.repoDs.delete(key)).errorOption:
-      trace "Error deleting block after failed quota update", err = err.msg
+  if res.kind == Stored:
+    trace "Block Stored"
+    if err =? (await self.updateQuotaUsage(plusUsed = res.used)).errorOption:
+      # rollback changes
+      without delRes =? await self.tryDeleteBlock(blk.cid), err:
```

Comment:
I will think about a way to address it, but really it's not that trivial. First of all we don't know if we should attempt to increase the quota unless we try to store a block - because it may turn out that the block is already stored and there was no reason to increase the quota in the first place.

It seems to be easy at surface. But there might occur concurrency errors, that eventually may lead to permanent inconsistencies in db, which should be avoided at all cost. Keep in mind that we need to keep consistent 4 different records in 2 different stores:
- block in repoDs
- block metadata in metaDs
- quota usage in metaDs
- total blocks in metaDs

Reply:
All of this should be written as a batch in one go with the `method put*(self: Datastore, batch: seq[BatchEntry])` method, this will avoid inconsistencies across this stores, the only exception being the `repoDs` which is right now a separate datastore. Another approach would be to moving the `repoDs` to the same table as the rest of the keys, but that might have additional performance consequences - but still worth a try.

We can also think about adding a concurrent store on top that will update/rollback a batch of `put` oprations in one transation.

Reply:
Following up on a offline discussion regarding this issue. We agreed that all the items we store in `metaDs`, which are:
* block metadata
* quota byte usage
* total number of blocks

should be saved with a single batch and a single transaction. However currently we don't have `modify` API that would accept batch input, and adding such API would require some additional development effort in `nim-datastore`, so we're fine with lower consistency guarantee, which is just consistency between block metadata
and the block data (a file).